File: commons/ts/reader.ts
import { parseArgs } from "util";

export class Reader {
  private readonly path: string;

  constructor() {
    const { values } = parseArgs({
      args: Bun.argv,
      options: {
        test: { type: "boolean", default: false },
      },
      strict: true,
      allowPositionals: true,
    });
    const fileName = values.test ? "sample.txt" : "data.txt";
    const parts = Bun.main.split("/");
    parts.reverse();
    const [year, day] = [parts[2], parts[1]];
    this.path = `data/${year}/${day}/${fileName}`;
  }

  async intLines(): Promise<number[]> {
    const lines = await this.lines();
    return lines.map((line) => parseInt(line));
  }

  async lines(): Promise<string[]> {
    const text = await this.read();
    return text.split("\n");
  }

  async read(): Promise<string> {
    return Bun.file(this.path).text();
  }
}

File: commons/elixir/reader.ex
defmodule Reader do
  @spec text() :: String.t()
  def text() do
    get_filepath() |> File.read!() |> String.trim()
  end

  @spec lines() :: [String.t()]
  def lines() do
    text() |> String.split("\n")
  end

  @spec get_filepath() :: Path.t()
  defp get_filepath() do
    {:current_stacktrace, frames} = Process.info(self(), :current_stacktrace)

    [year, day, _] =
      frames
      |> Enum.map(fn {_, _, _, meta} -> meta[:file] end)
      |> Enum.find(fn path -> Path.basename(path) == "solver.ex" end)
      |> Path.split()

    test? = Enum.member?(System.argv(), "--test")
    file = if test?, do: "sample.txt", else: "data.txt"

    Path.join(["data", year, day, file])
  end
end
